xds/internal/xdsclient/authority.go

```diff
@@ -306,365 +334,281 @@ func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Ty
+		a.handleADSResourceDoesNotExist(rType, resourceName)
+	})
 }
 
-// newConnectionError is called by the underlying transport when it receives a
-// connection error. The error will be forwarded to all the resource watchers.
-func (a *authority) newConnectionError(err error) {
-	a.resourcesMu.Lock()
-	defer a.resourcesMu.Unlock()
-
-	a.stopWatchTimersLocked()
+func (a *authority) handleADSResourceDoesNotExist(rType xdsresource.Type, resourceName string) {
+	if a.logger.V(2) {
+		a.logger.Infof("Watch for resource %q of type %s timed out", resourceName, rType.TypeName())
+	}
 
-	// We do not consider it an error if the ADS stream was closed after having received
-	// a response on the stream. This is because there are legitimate reasons why the server
-	// may need to close the stream during normal operations, such as needing to rebalance
-	// load or the underlying connection hitting its max connection age limit.
-	// See gRFC A57 for more details.
-	if xdsresource.ErrType(err) == xdsresource.ErrTypeStreamFailedAfterRecv {
-		a.logger.Warningf("Watchers not notified since ADS stream failed after having received at least one response: %v", err)
+	resourceStates := a.resources[rType]
+	if resourceStates == nil {
+		if a.logger.V(2) {
+			a.logger.Infof("Resource %q of type %s currently not being watched", resourceName, rType.TypeName())
+		}
 		return
 	}
-
-	for _, rType := range a.resources {
-		for _, state := range rType {
-			// Propagate the connection error from the transport layer to all watchers.
-			for watcher := range state.watchers {
-				watcher := watcher
-				a.serializer.TrySchedule(func(context.Context) {
-					watcher.OnError(xdsresource.NewErrorf(xdsresource.ErrorTypeConnection, "xds: error received from xDS stream: %v", err), func() {})
-				})
-			}
+	state, ok := resourceStates[resourceName]
+	if !ok {
+		if a.logger.V(2) {
+			a.logger.Infof("Resource %q of type %s currently not being watched", resourceName, rType.TypeName())
 		}
+		return
 	}
-}
 
-// Increments the reference count. Caller must hold parent's authorityMu.
-func (a *authority) refLocked() {
-	a.refCount++
-}
-
-// Decrements the reference count. Caller must hold parent's authorityMu.
-func (a *authority) unrefLocked() int {
-	a.refCount--
-	return a.refCount
-}
-
-func (a *authority) close() {
-	a.transport.Close()
-
-	a.resourcesMu.Lock()
-	a.closed = true
-	a.resourcesMu.Unlock()
+	state.cache = nil
+	state.md = xdsresource.UpdateMetadata{Status: xdsresource.ServiceStatusNotExist}
+	for watcher := range state.watchers {
+		watcher := watcher
+		a.watcherCallbackSerializer.TrySchedule(func(context.Context) { watcher.OnResourceDoesNotExist(func() {}) })
+	}
 }
 
+// watchResource registers a new watcher for the specified resource type and
+// name. It returns a function that can be called to cancel the watch.
+//
+// If this is the first watch for any resource on this authority, an xdsChannel
+// to the management server will be created.
```

Comment:
Nit: is there a way to tersely explain what management server this will connect to. "active"? "First server".

Reply:
sg